Jpeg HuffmanTable improvements

br3aker opened the pull request:
### Prerequisites

- [x] I have written a descriptive pull-request title
- [x] I have verified that there are no overlapping [pull-requests](https://github.com/SixLabors/ImageSharp/pulls) open
- [x] I have verified that I am following the existing coding patterns and practice as demonstrated in the repository. These follow strict Stylecop rules :cop:.
- [x] I have provided test coverage for my change (where applicable)

### Description
This PR removes redundant `Configure()` calls for `HuffmanTable` and removes stack pressure from `stackalloc` calls.

### TODO
- [x] Remove `Configure()` method 
- [x] Remove `stackalloc` calls
- [x] Create out-of-bounds huffman table test image

JimBobSquarePants commented on src/ImageSharp/Formats/Jpeg/Components/Decoder/HuffmanTable.cs:
```diff
@@ -62,20 +55,7 @@ internal unsafe struct HuffmanTable
         public HuffmanTable(ReadOnlySpan<byte> codeLengths, ReadOnlySpan<byte> values)
         {
-            this.isConfigured = false;
-            Unsafe.CopyBlockUnaligned(ref this.Sizes[0], ref MemoryMarshal.GetReference(codeLengths), (uint)codeLengths.Length);
```
As I recall this copy exists because the input span is allocated once and reused for each Huffman table.

br3aker replied:
Which one? `codeLengths`? It may be unique for each table. Right now ctor copies input spans and then builds actual table representation during `Configure()` call - that's the only reason it copies first span param to internal array. While calling this method a couple of times with short-circuit `return` is not bad, it's impossible to pass workspace spans from allocator this way. That's why I want to build table in the ctor but that remote executor crash...

I hope it's okay if I can use github actions for some testing. That random crash is very strange and I'm not sure I can debug it locally.

JimBobSquarePants replied:
Ah yes, The reason for the lazy invocation was to avoid building the tables to allow the validation of successive scan parameters for progressive images. See `HuffmanScanDecoder.CheckProgressiveData`. This is the same thing that libjpeg and libjpeg-turbo does. https://github.com/libjpeg-turbo/libjpeg-turbo/blob/main/jdhuff.c#L77

JimBobSquarePants replied:
Also yes, experiment away. Maybe disable the codecov run first though to save time.

br3aker replied:
Hm, that check seems valid but table building is not that expensive, especially compared to exception path imo. Should I still remove `Configure` and move everything to the ctor for simplicity and reduced `stackalloc`s?

JimBobSquarePants replied:
Let’s give it a go yeah!

br3aker commented:
Gotcha

```
Message: 
    Assert.Throws() Failure
    Expected: typeof(SixLabors.ImageSharp.InvalidImageContentException)
    Actual:   typeof(System.IndexOutOfRangeException): Index was outside the bounds of the array.
    ---- System.IndexOutOfRangeException : Index was outside the bounds of the array.

  Stack Trace: 
    HuffmanTable.ctor(ReadOnlySpan`1 codeLengths, ReadOnlySpan`1 values) line 58
    JpegDecoderCore.ProcessDefineHuffmanTablesMarker(BufferedReadStream stream, Int32 remaining) line 1098
    JpegDecoderCore.ParseStream(BufferedReadStream stream, HuffmanScanDecoder scanDecoder, CancellationToken cancellationToken) line 358
    ...
```

One of the invalid test images was going out of bounds during new `HuffmanTable` ctor. The only reason it wasn't spotted before is that exposed code in `Configure()` was called after checks for invalid image which led to this crashes.

JimBobSquarePants commented:
Ah nice, well done finding that. So now I guess we have to find a way to sanitise cheaply and then try to move the table building to the constructor.

br3aker commented:
It's actually a bug, there's a possibility for malformed image go out of bounds without any exceptions in current master. I found a way to check that without a noticable slowdown. I'd also create an example image for a test suite.

br3aker commented on src/ImageSharp/Formats/Jpeg/Components/Decoder/HuffmanTable.cs:
```diff
@@ -94,6 +94,17 @@ public HuffmanTable(ReadOnlySpan<byte> codeLengths, ReadOnlySpan<byte> values)
+                    // a new code which occupies one extra bit, e.g. 1111(1) +1 => (1)1111(0)
+                    if (code >= (1 << si))
+                    {
+                        JpegThrowHelper.ThrowInvalidImageContentException("Bad huffman table.");
```
@JimBobSquarePants do need a more specific exception message for this?

JimBobSquarePants replied:
No, that should be fine

br3aker commented:
High hopes it would pass all the tests, it is ready to be reviewed:

- Rewritten part of the huffman table building routine, it now requires 256 bytes less memory workspace
- No `stackalloc` calls
- Rewritten DHT marker parsing code, now it fetches a big single buffer for binary parsing and table building routine
- Added sanity checks for huffman table input, now it should be impossible to break decoder with malformed DHT marker

There's no need for a new test image, image from [#839](https://github.com/SixLabors/ImageSharp/issues/839) has malformed DHT marker which goes out-of-bounds during huffman table for unsafe pointers (which is unspottable without wrapping them with `Span`). I've left a note just in case but this PR should nullify this error.

*this is my favourite*
On top of a critical security breach fix and extra checks we are not  actually slower:

```
MASTER
// Decoding
// Elapsed: 4643ms across 400 iterations
// Average: 11,6075ms

PR
// Decoding
// Elapsed: 4634ms across 400 iterations
// Average: 11,585ms
```

JimBobSquarePants reviewed (commented):
Faster... Fantastic 🤣

This is very, very good as always. Thanks! 👍
